### PR TITLE
[#885] make LSP server exit behaviour configurable

### PR DESCRIPTION
**Pull request by cdietrich**
Signed-off-by: Christian Dietrich <christian.dietrich@itemis.de>

### CODE REVIEW

**Comment by cdietrich** [comment]
@kaikreuzer is this what you were thinking of?
@svenefftinge please review

**Comment by kaikreuzer** [comment]
Yes, thanks, that should do the trick!
Just one question: If I am not mistaken, the termination is a two step process: shutdown & exit.
So would we need a similar mechanism for the shutdown step as well? As mentioned [here](https://github.com/eclipse/smarthome/issues/6291#issuecomment-425843106), I once had an issue to reconnect to the server after avoiding the exit call as it was replying that it is currently shutting down and won't allow any connections.

**Comment by cdietrich** [comment]
since i dont have your usecase we would need more detailed input on that.

**Comment by kaikreuzer** [comment]
I was hoping that you knew what exactly the LSP server is doing when receiving the [shutdown request](https://microsoft.github.io/language-server-protocol/specification#shutdown) and whether this leaves it in an unusable state or not.

**Comment by cdietrich** [comment]
i have heavy doubts it leaves a recoverable states. it is a noop. (just sets a boolean)

**Comment by kaikreuzer** [comment]
> i have heavy doubts it leaves a recoverable state

You mean you are pretty sure it leaves a recoverable state as nothing really happens upon such a request? Then all should be fine.

Unfortunately, I didn't capture the exact exception that I had seen at that time - maybe it didn't say that "shutdown has been called already", but that ["the server has already been initialized"](https://github.com/eclipse/xtext-core/blob/ceca3728be5b57e33a6727509128a482a483f0b1/org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/LanguageServerImpl.xtend#L138)? I'm not sure, just wanted to mention it that the re-connection should be tested.

**Comment by cdietrich** [comment]
no i mean: shutdown is i a noop. code is not tested with other scenarios.
=> we require your testing/input

**Comment by kaikreuzer** [comment]
> i am 100% sure a new try will fail

Why?

Since I merged my workaround with https://github.com/openhab/openhab-core/pull/409, I didn't get any negative feedback from the users - it all seems to work smoothly. So having shutdown a noop and exit ignored seems to be ok.

**Comment by cdietrich** [comment]
i am not sure if e.g. the  shutdowncalled flag is properly reset

**Comment by cdietrich** [comment]
and it will explode if you call initialize twice

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerExitHandler.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+ * 
+ * @since 2.16
+ */
+@ImplementedBy(ILanguageServerExitHandler.DefaultImpl.class)
+public interface ILanguageServerExitHandler {
+
+	/**
+	 * Callback that is called when Language Servers <code>exit</code> is called.
+	 */
+	void exit(int status);
```
Comment (svenefftinge) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerExitHandler.java]: Shouldn't we also add the `shutdown` method here and remove the argument?

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/ILanguageServerExitHandler.java]: will have a look

**Comment by cdietrich** [comment]
@svenefftinge updated the PR. please review again

**Comment by svenefftinge** [comment]
Ideally the hook would not be provided through DI but per configuration, so that when using the launcher we would not need to call `System.exit` but would simply stop listening and naturally terminate.

**Comment by cdietrich** [comment]
but wont that be then up to LSP4j and the launchers there and default impl in the shutdown and exit methods?

**Pull request merged**